Add a best practice page for tuning remote write

csmarchbanks opened the pull request:
Documentation for what each remote write parameter does and when to adjust them. Any feedback is appreciated!

Closes: https://github.com/prometheus/prometheus/issues/5803

cc: @beorn7 @cstyan

beorn7 commented:
Many thanks. @juliusv could you have a look as well, as an authoritative -team reviewer? You are, after all, the maintainer of the remote storage parts.

juliusv commented on content/docs/practices/index.md:
```diff
@@ -1,5 +1,5 @@
 ---
 title: Best practices
-sort_rank: 8
+sort_rank: 9
```
I think this needs to stay the same, as this is about the sort order of the "Best practices" menu item among the top-level items.

juliusv commented on content/docs/practices/remote_write.md:
```diff
@@ -0,0 +1,103 @@
+## Remote write characteristics
+
+Each remote write destination starts a queue which reads from the write-ahead
+log (WAL), writes the samples into a queue owned by a shard, which then sends a
```
Might be good to clarify that these queues are in-memory only ("writes the samples into a queue" sounds like it's on-disk queues). At least that's how I assume it still works.

juliusv commented on content/docs/practices/remote_write.md:
```diff
@@ -0,0 +1,103 @@
+the remote endpoint remains down for more than 2 hours. After 2 hours, the WAL
+will be compacted and data that has not been sent will be lost.
+
+During operation, Prometheus will continuously calculate the the optimal number
```
"the the" -> "the"

juliusv commented on content/docs/practices/remote_write.md:
```diff
@@ -0,0 +1,103 @@
+report ~25% increased memory usage, but that number is dependent on the shape
+of the data. A mapping of series id to label values is stored for each series
+in the WAL, so large amounts of series churn significantly increases memory
+usage.
```
Might be good to clarify how and which portions of the on-disk WAL are mapped / copied into memory, to give more context here.

csmarchbanks replied:
This paragraph tries to specify that it is a mapping of series id to label values, do you have any suggestions for how to clarify that?

juliusv replied:
Ahh, I parenthesized sentence fragments together incorrectly in my mind: that for each series, there is a mapping cached in the WAL, not that for each series in the WAL, there is a mapping cached (in memory).

How about (to make the structure extra clear): "For each series in the WAL, the remote write code caches a mapping of series ID to label values."

csmarchbanks replied:
:+1: That sounds good. I have made the change.

juliusv commented on content/docs/practices/remote_write.md:
```diff
@@ -0,0 +1,103 @@
+Min shards configures the minimum number of shards used by Prometheus, and is
+the number of shards used when remote write starts. If remote write falls
+behind, Prometheus will automatically scale up the number of shards so most
+users do not have to adjust this parameter. Setting to a higher number will
```
Setting -> Setting this

juliusv commented on content/docs/practices/remote_write.md:
```diff
@@ -0,0 +1,103 @@
+---
+title: Remote Write Tuning
```
We've standardized on not using title case for menu items, so:

"Remote Write Tuning" -> "Remote write tuning"

juliusv commented:
:+1: besides comments.

I haven't even followed any of the recent remote write developments closely, so either @tomwilkie or @csmarchbanks should probably take over maintainership.

beorn7 commented:
@cstyan might also be a candidate. Let's ask around who wants the job. :o)

csmarchbanks commented:
I think I have addressed all the comments at this point, let me know if I missed anything.

As far as maintainership goes, I think all of the people mentioned would be good candidates. I would be happy to take over remote write at some point; it is an area that I am familiar with and use heavily. But I would support @cstyan or @tomwilkie as well.

juliusv commented:
:+1: from my side

beorn7 commented on content/docs/practices/remote_write.md:
```diff
@@ -0,0 +1,103 @@
+Max shards configures the maximum number of shards, or parallelism, Prometheus
+will use for each remote write queue. Prometheus will try not to use too many
+shards, but if the queue falls behind the remote write component will increase
+the number of shards up to max shards to increase thoughput.
```
Perhaps add something about the memory impact (as the number of shardes multiplies with the capacity to determine how many samples are queued at max). Also, [there were reports of lock contention with too many shards](https://github.com/prometheus/prometheus/issues/5803#issuecomment-520973662). Perhaps give some sane ballpark numbers here so that people don't set this to thousands of shards?

csmarchbanks replied:
I added more information about the memory impact of shards. I did not add anything specific about lock contention, but did say that it is unlikely to need to increase the default value for `max_shards`.

beorn7 commented:
Thank you very much. Let's merge this now. We can still add more if that appears necessary in the future.
